### DISCUSSION TITLE
Grain tracker with ebsd reader and nucleation causes error

### DISCUSSION THREAD

**Discussion by FHilty** [idaholab/moose#20198] [Q&A Modules: Phase field] [answered]
Hello Everyone,

I mentioned this problem in another discussion about the grain tracker, but it seemed divergent from the issue in that thread. I want to load a grain structure with the ebsd reader then model dynamic recrystallization using the Grain Tracker and Discrete Nucleation. Everything seems to progress fine until a nucleus meets the requirements to be remapped with the grain tracker, then an error is produced from the ebsd reader:

> *** ERROR ***
The following error occurred in the object "ebsd_reader", of type "EBSDReader".
>
>Error! Index out of range in EBSDReader::indexFromIndex(), index: 10 size: 9

The system starts with 9 grains, adding a 10th grain breaks breaks the ebsd reader. Attached is a reduced input file based on the grain_tracker_ebsd test in the phase field module and the ebsd file from that test, which will produce the error. Hopefully combining these features is possible and I am just missing something, any help would be appreciated.

Floyd

[grain_tracker_ebsd_nucleation_error.txt](https://github.com/idaholab/moose/files/7982474/grain_tracker_ebsd_nucleation_error.txt)
[ebsd_9.txt](https://github.com/idaholab/moose/files/7982477/ebsd_9.txt)

**Comment by chaitanyaBhave26** [comment, marked as answer]
@FHilty can you try setting 

  [ebsd_reader]
    type = EBSDReader
    **execute_on = INITIAL**
  []
   [ebsd]
    type = PolycrystalEBSD
    coloring_algorithm = jp
    ebsd_reader = ebsd_reader
   **execute_on = INITIAL**
  []
 
The ebsd reader should not be running past the zeroth timestep, I think that is what is causing your issue.

**Reply by ykvishal** [reply]
When I commented following section then your input file is working

  ```
[phi1]
    type = OutputEulerAngles
    euler_angle_provider = ebsd_reader
    output_euler_angle = phi1
    grain_tracker = grain_tracker
    variable = phi1
  []
  [Phi]
    type = OutputEulerAngles
    euler_angle_provider = ebsd_reader
    output_euler_angle = Phi
    grain_tracker = grain_tracker
    variable = Phi
  []
  [phi2]
    type = OutputEulerAngles
    euler_angle_provider = ebsd_reader
    output_euler_angle = phi2
    grain_tracker = grain_tracker
    variable = phi2
  []
```



Here is the animation video:

https://user-images.githubusercontent.com/31565781/152074479-afbe8ee5-b037-424f-883e-0f296dd1221f.mP4

**Reply by FHilty** [reply]
That was indeed the culprit, thank you very much Vishal!

Floyd

**Reply by ykvishal** [reply]
@FHilty If you are planning to perform large scale simulations then you should use @chaitanyaBhave26 ' s modified EBSDReader, which is more memory efficient compared to the current EBSDReader.  Otherwise, current EBSDReader is good for small system size and number of grains.

**Reply by FHilty** [reply]
In the near term the systems will be small, but i expect to have to go up to 100+ grains eventually. I will keep that in mind, thank you!
